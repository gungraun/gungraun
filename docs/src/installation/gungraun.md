# Gungraun

Gungraun is divided into the library `gungraun` and the benchmark runner
`gungraun-runner`.

## Installation of the library

To start with Gungraun, add the following to your `Cargo.toml` file:

```toml
[dev-dependencies]
gungraun = "0.16.1"
```

or run

```bash
cargo add --dev gungraun@0.16.1
```

## Installation of the benchmark runner

To be able to run the benchmarks you'll also need the `gungraun-runner`
binary installed somewhere in your `$PATH`. Otherwise, there is no need to
interact with `gungraun-runner` as it is just an implementation detail.

### From Source

```shell
cargo install --version 0.16.1 gungraun-runner
```

There's also the possibility to install the binary somewhere else and point the
`GUNGRAUN_RUNNER` environment variable to the absolute path of the
`gungraun-runner` binary like so:

```shell
cargo install --version 0.16.1 --root /tmp gungraun-runner
GUNGRAUN_RUNNER=/tmp/bin/gungraun-runner cargo bench --bench my-bench
```

### Binstall

The `gungraun-runner` binary is
[pre-built](https://github.com/gungraun/gungraun/releases/tag/v0.16.1)
for most platforms supported by valgrind and easily installable with
[binstall](https://github.com/cargo-bins/cargo-binstall)

```shell
cargo binstall gungraun-runner@0.16.1
```

## Updating

When updating the `gungraun` library, you'll also need to update
`gungraun-runner` and vice-versa or else the benchmark runner will exit
with an error.

### In the Github CI

Since the `gungraun-runner` version must match the `gungraun` library
version it's best to automate this step in the CI. A job step in the github
actions CI could look like this

```yaml
- name: Install gungraun-runner
  run: |
    version=$(cargo metadata --format-version=1 |\
      jq '.packages[] | select(.name == "gungraun").version' |\
      tr -d '"'
    )
    cargo install gungraun-runner --version $version
```

Or, speed up the overall installation time with `binstall` using the
[taiki-e/install-action](https://github.com/taiki-e/install-action)

```yaml
- uses: taiki-e/install-action@cargo-binstall
- name: Install gungraun-runner
  run: |
    version=$(cargo metadata --format-version=1 |\
      jq '.packages[] | select(.name == "gungraun").version' |\
      tr -d '"'
    )
    cargo binstall --no-confirm gungraun-runner --version $version
```
