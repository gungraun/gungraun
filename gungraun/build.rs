//! The build script

// spell-checker: ignore rustified iquote iquotevalgrind

#[cfg(feature = "client_requests_defs")]
mod imp {
    use std::borrow::Cow;
    use std::fmt::Display;
    use std::io::{BufRead, BufReader, Cursor};
    use std::path::PathBuf;

    use bindgen::{builder, Bindings};
    use rustc_version::{version, Version};
    use strum::{EnumIter, IntoEnumIterator};

    #[derive(Debug)]
    struct Target {
        arch: String,
        env: String,
        os: String,
        triple: String,
        vendor: String,
    }

    #[derive(EnumIter, Debug, PartialEq, Eq)]
    enum Support {
        Arm,
        Aarch64,
        X86,
        X86_64,
        Riscv64,
        Native,
        No,
    }

    impl Display for Support {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            let support = format!("{self:?}").to_lowercase();
            f.write_str(&support)
        }
    }

    impl Target {
        fn from_env() -> Self {
            Self {
                arch: std::env::var("CARGO_CFG_TARGET_ARCH").unwrap(),
                env: std::env::var("CARGO_CFG_TARGET_ENV").unwrap(),
                os: std::env::var("CARGO_CFG_TARGET_OS").unwrap(),
                vendor: std::env::var("CARGO_CFG_TARGET_VENDOR").unwrap(),
                triple: std::env::var("TARGET").unwrap(),
            }
        }
    }

    fn print_client_requests_support(value: &Support) {
        println!("cargo:rustc-cfg=client_requests_support=\"{value}\"");
    }

    fn include_dirs(target: &Target) -> impl Iterator<Item = String> {
        [
            Cow::Owned(format!(
                "IAI_CALLGRIND_{}_VALGRIND_INCLUDE",
                target.triple.replace('-', "_").to_ascii_uppercase()
            )),
            Cow::Borrowed("IAI_CALLGRIND_VALGRIND_INCLUDE"),
        ]
        .into_iter()
        .filter_map(|env| std::env::var(env.as_ref()).ok())
    }

    fn build_native(target: &Target) {
        let mut builder = cc::Build::new();

        for env in include_dirs(target) {
            builder.include(env);
        }

        if target.os == "freebsd" {
            builder.include("/usr/local/include");
        }

        if let Ok(env) = std::env::var("GUNGRAUN_CROSS_TARGET") {
            let path = PathBuf::from("/valgrind/target/valgrind")
                .join(env)
                .join("include");
            builder.include(path);
        }

        builder.include("valgrind/include");

        builder
            .debug(true)
            .file("valgrind/native.c")
            .compile("native");
    }

    fn build_bindings(target: &Target) -> Bindings {
        let mut builder = builder();

        for env in include_dirs(target) {
            builder = builder.clang_arg(format!("-iquote{env}"));
        }

        if let Ok(env) = std::env::var("GUNGRAUN_CROSS_TARGET") {
            let path = PathBuf::from("/valgrind/target/valgrind")
                .join(env)
                .join("include");
            builder = builder.clang_arg(format!("-iquote{}", path.display()));
        }

        if target.os == "freebsd" {
            builder = builder.clang_arg("-iquote/usr/local/include");
        }

        let bindings = builder
            .clang_arg("-iquote/usr/include")
            .clang_arg("-iquotevalgrind/include")
            .header("valgrind/wrapper.h")
            .allowlist_var("IC_IS_PLATFORM_SUPPORTED_BY_VALGRIND")
            .allowlist_var("IC_VALGRIND_MAJOR")
            .allowlist_var("IC_VALGRIND_MINOR")
            .allowlist_type("IC_.*ClientRequest")
            .rustified_enum("IC_.*ClientRequest")
            .layout_tests(false)
            .parse_callbacks(Box::new(bindgen::CargoCallbacks::new()))
            .generate()
            .expect("Generating binding should succeed");

        let out_dir = std::env::var("OUT_DIR").map(PathBuf::from).unwrap();
        let path = out_dir.join("bindings.rs");
        bindings.write_to_file(path).unwrap();
        bindings
    }

    // Return the rust version if running rustc was successful
    fn get_rust_version() -> Option<Version> {
        version().ok()
    }

    pub fn main() {
        println!("cargo:rerun-if-changed=valgrind/wrapper.h");
        println!("cargo:rerun-if-changed=valgrind/native.c");
        println!("cargo:rerun-if-env-changed=IAI_CALLGRIND_VALGRIND_INCLUDE");
        println!("cargo:rerun-if-env-changed=GUNGRAUN_CROSS_TARGET");
        println!("cargo:rerun-if-env-changed=TARGET");

        // rustc-check-cfg is introduced in rust with version 1.80 and avoids the compiler warnings
        // in version >= 1.80.0. Printing it when compiling with versions < 1.80 triggers a warning,
        // too. To get the best of both worlds we check against the currently active rust version.
        if let Some(rust_version) = get_rust_version() {
            if rust_version.major >= 1 && rust_version.minor >= 80 {
                let values = Support::iter()
                    .map(|s| format!("\"{s}\""))
                    .collect::<Vec<String>>()
                    .join(",");
                println!("cargo:rustc-check-cfg=cfg(client_requests_support,values({values}))");
            }
        }

        let target = Target::from_env();

        // When building the docs on docs.rs we can take a shortcut
        if std::env::var("DOCS_RS").is_ok() {
            print_client_requests_support(&Support::X86_64);
            build_bindings(&target);
            build_native(&target);
            return;
        }

        let bindings = build_bindings(&target);

        // These guards mirror the checks in the `valgrind.h` header file
        let support = if target.arch == "x86_64"
            && (target.os == "linux"
                || target.os == "freebsd"
                || (target.vendor == "apple" && target.os == "darwin")
                || (target.os == "windows" && target.env == "gnu")
                || ((target.vendor == "sun") || target.vendor == "pc") && target.os == "solaris")
        {
            Some(Support::X86_64)
        } else if target.arch == "x86"
            && (target.os == "linux"
                || target.os == "freebsd"
                || (target.vendor == "apple" && target.os == "darwin")
                || (target.os == "windows" && target.env == "gnu")
                || ((target.vendor == "sun") || target.vendor == "pc") && target.os == "solaris")
        {
            Some(Support::X86)
        } else if target.arch == "arm" && target.os == "linux" && target.env == "gnu" {
            Some(Support::Arm)
        } else if target.arch == "aarch64"
            && (target.os == "freebsd" || (target.os == "linux" && target.env == "gnu"))
        {
            Some(Support::Aarch64)
        } else if target.arch == "riscv64gc" && target.os == "linux" {
            Some(Support::Riscv64)
        } else {
            let re = regex::Regex::new(
                r"IC_IS_PLATFORM_SUPPORTED_BY_VALGRIND.*?=\s*(?<value>true|false)",
            )
            .expect("Regex should compile");
            let reader = BufReader::new(Cursor::new(bindings.to_string()));
            let mut support = None;
            for line in reader.lines().map(Result::unwrap) {
                if let Some(caps) = re.captures(&line) {
                    let value = caps.name("value").unwrap().as_str();
                    if value == "false" {
                        support = Some(Support::No);
                    } else if value == "true" {
                        support = Some(Support::Native);
                    } else {
                        // do nothing
                    }
                    break;
                }
            }
            support
        };

        if let Some(support) = support {
            print_client_requests_support(&support);
            if support != Support::No {
                build_native(&target);
            }
        } else {
            eprintln!("{bindings}");
            panic!("Unable to set cfg value for client_requests_support");
        }
    }
}

#[cfg(not(feature = "client_requests_defs"))]
mod imp {
    pub fn main() {}
}

fn main() {
    imp::main();
}
