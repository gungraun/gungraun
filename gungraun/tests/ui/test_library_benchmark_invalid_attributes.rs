use gungraun::library_benchmark;

#[library_benchmark]
#[b]
fn bench1() {}

#[library_benchmark]
#[inline(never)]
fn bench2() {}

#[inline(never)]
#[library_benchmark]
fn bench3() {}

#[bench::id()]
#[library_benchmark]
fn bench4() {}

fn main() {}
