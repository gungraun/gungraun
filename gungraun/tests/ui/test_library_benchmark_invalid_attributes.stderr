error: Invalid attribute: 'b'

         = help: Only the `bench` and the `benches` attribute are allowed
         = note: #[bench::my_id("with", "args")]
                           or #[benches::my_id(args = [("with", "args"), ...])]

 --> tests/ui/test_library_benchmark_invalid_attributes.rs:4:1
  |
4 | #[b]
  | ^^^^

error: Invalid attribute: 'inline'

         = help: Only the `bench` and the `benches` attribute are allowed
         = note: #[bench::my_id("with", "args")]
                           or #[benches::my_id(args = [("with", "args"), ...])]

 --> tests/ui/test_library_benchmark_invalid_attributes.rs:8:1
  |
8 | #[inline(never)]
  | ^^^^^^^^^^^^^^^^

error: Invalid attribute: 'inline'

         = help: Only the `bench` and the `benches` attribute are allowed
         = note: #[bench::my_id("with", "args")]
                           or #[benches::my_id(args = [("with", "args"), ...])]

  --> tests/ui/test_library_benchmark_invalid_attributes.rs:11:1
   |
11 | #[inline(never)]
   | ^^^^^^^^^^^^^^^^

error[E0433]: failed to resolve: use of undeclared crate or module `bench`
  --> tests/ui/test_library_benchmark_invalid_attributes.rs:15:3
   |
15 | #[bench::id()]
   |   ^^^^^ use of undeclared crate or module `bench`
